### DISCUSSION TITLE
A question about stiffness matrix, assembly and assembly order in MOOSE

### DISCUSSION THREAD

**Discussion by alimostafavi24** [idaholab/moose#28626] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Dear MOOSE experts,

I am writing this to inquire about stiffness matrix, assembly and assembly order in the framework.

First, is there any documentation regarding the stiffness matrix (or LHS) and RHS in Navier-Stokes and Phase-Field modules?

Second, if someone couples the two, the framework forms the assembled matrix. But what would be the order? (Variables are velocity vector, pressure, order parameter and chemical potential). -mat_view gives the matrix data, but how about the order of nonlinear variables? Is there anyway to see the right-hand-side of the equations too or we can just see the stiffness matrix?

Thanks for your assistance.

Regards,

**Comment by GiudGiud** [comment]
Hello

> First, is there any documentation regarding the stiffness matrix (or LHS) and RHS in Navier-Stokes and Phase-Field modules?

Yes in the sense that all the kernels are documented for Navier Stokes, and many of them for phase field (but it's not complete yet there). 
This tells you how we assemble the residual and Jacobian.

In most solves (linear finite volume being the exception I know), we dont form the RHS explicitly. If you want to create it, you can use the tagging system on each kernel that should contribute to it
https://mooseframework.inl.gov/source/interfaces/TaggingInterface.html

Guillaume

**Reply by alimostafavi24** [reply]
Ok. I had a question regarding obtaining the stiffness matrix. Considering solving the transient implicit Phase-field in Split form (c order parameter, w chemical potential), -mat_view returns this matrix (it is the sparsity pattern). I reorder it in this way:[k_cc k_cw;k_wc k_ww][c;w]=[F_1;F_2]. 

What exactly this matrix could be? I am asking this because stiffness matrix sdoes not contain time derivative terms. Am I right? But the figure has the dc/dt term contributions. Can this output be kind of sum of the mass and stiffenss matrix?
![K](https://github.com/user-attachments/assets/7422b867-4421-41e9-bc95-bb18a7ca4e0e)

**Reply by GiudGiud** [reply]
Sorry forgot to reply. 

Yes this is the Jacobian not the stiffness matrix. You would need to remove the time derivative terms at least.
